### PR TITLE
CLID-612: Move integration tests within oc-mirror

### PR DESCRIPTION
**Pull request by adolfo-ab**
# Description

This PR moves the integration tests from https://github.com/oc-mirror-test/integration to oc-mirror itself.
The integration tests are a separate module within oc-mirror.
It removes the artifacts image and it also changes the registry that we use in the tests to an in-process registry.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code Improvements (Refactoring, Performance, CI upgrades, etc)
- [ ] Internal repo assets (diagrams / docs on github repo)
- [ ] This change requires a documentation update on openshift docs

# How Has This Been Tested?
Running the tests directly and from their container.

## Expected Outcome
All the tests pass.

## Summary by CodeRabbit

* **Tests**
  * Added a comprehensive integration test suite: end-to-end mirroring, disk<->mirror flows, delete workflows, dry-run and exit-code scenarios, archive edge cases, operator/catalog listings, builders, helpers, runner/registry test tooling, fixtures and test data.
* **Chores**
  * Added a Make target to run integration tests with verbose, single-run behavior and an extended timeout.

### CODE REVIEW

File: tests/integration/pkg/registry/registry.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+// Registry manages an in-process distribution/distribution registry instance.
+type Registry struct {
+	service     *dregistry.Registry
+	storagePath string
+	port        int
+	client      name.Registry
+}
+
+// Start parses the config file, creates an in-process registry, and starts it in the background.
+func Start(ctx context.Context, configPath string, port int, logWriter ...interface{}) (*Registry, error) {
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/pkg/registry/registry.go]: Fixed

File: tests/integration/pkg/registry/registry.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to read registry config: %w", err)
+	}
+
+	config, err := configuration.Parse(bytes.NewReader(configData))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse registry config: %w", err)
+	}
+
+	storagePath := filepath.Join(config.Storage["filesystem"]["rootdirectory"].(string), "docker")
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/pkg/registry/registry.go]: Fixed

File: tests/integration/pkg/registry/registry.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		return nil, fmt.Errorf("failed to set up registry client: %w", err)
+	}
+
+	return &Registry{
+		service:     service,
+		storagePath: storagePath,
+		port:        port,
+		client:      client,
+	}, nil
+}
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/pkg/registry/registry.go:81]: Fixed

File: tests/integration/testdata/registry-config.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    service: registry
+storage:
+  delete:
+      enabled: true
+  cache:
+    blobdescriptor: inmemory
+  filesystem:
+    rootdirectory: /tmp/
+http:
+  addr: :5000
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/testdata/registry-config.yaml:16]: Fixed

File: tests/integration/d2m_invalid_from_test.go [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+			"--remove-signatures=true", "--dest-tls-verify=false")
+		expectOcMirrorCommandInvalidFrom(result, err)
+	})
+})
+
+func expectOcMirrorCommandInvalidFrom(result *ocmirror.Result, err error) {
+	Expect(err).ToNot(HaveOccurred())
+	Expect(result.ExitCode).ToNot(BeZero(), "should be non-zero exit status")
+	Expect(result.Stdout).To(ContainSubstring(`no tar archives matching "mirror_[0-9]{6}\\.tar"`))
+}
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/d2m_invalid_from_test.go:47]: This comment is not correct.

File: tests/integration/exit_codes_test.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+		isc := filepath.Join(iscDir, "exit_codes", "isc-additional-image-error.yaml")
+
+		By("running M2D with only nonexistent additional images")
+		result, err := runner.MirrorToDisk(ctx, isc, workDir)
+		expectOcMirrorExitCode(result, err, 8,
+			"some errors occurred during the mirroring",
+			"registry.invalid.example.com/nonexistent/image:latest",
+			"quay.io/nonexistent-org/nonexistent-image:v1.0.0",
+		)
+	})
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/exit_codes_test.go:63]: Redundant assertion and out of the scope of this PR

File: tests/integration/go.mod [hunk 7]
```diff
@@ -0,0 +1,134 @@
+module github.com/openshift/oc-mirror/tests/integration
+
+go 1.24.3
+
+require (
+	github.com/distribution/distribution/v3 v3.0.0
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/go.mod]: Fixed

File: tests/integration/list_test.go [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+var _ = Describe("oc-mirror list releases", Ordered, func() {
+	// Initialize oc-mirror runner
+	binaryPath := os.Getenv("OC_MIRROR_BINARY")
+	// Custom runner so we can override envs without affecting other tests
+	releaseRunner := ocmirror.NewRunner(binaryPath)
+	BeforeAll(func() {
+		if skipListTest() {
+			Skip("oc-mirror list operators not available in < 4.22")
+		}
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/list_test.go:70]: Fixed

File: tests/integration/helpers_test.go [hunk 9]
```diff
@@ -0,0 +1,792 @@
...
+	Expect(parts).ToNot(BeNil(), "invalid version format %q", ver)
+	verMajor, err := strconv.Atoi(parts[1])
+	Expect(err).ToNot(HaveOccurred(), "failed to parse oc mirror major version %q", ver)
+	if verMajor == major {
+		verMinor, err := strconv.Atoi(parts[2])
+		Expect(err).ToNot(HaveOccurred(), "failed to parse oc mirror minor version %q", ver)
+		return verMinor < minor
+	}
+	return verMajor < major
+}
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/helpers_test.go:793]: Out of scope for this PR

File: tests/integration/helpers_test.go [hunk 10]
```diff
@@ -0,0 +1,792 @@
...
+	idmsFileName        = "idms-oc-mirror.yaml"
+
+	// cacheRepositoriesSubdir is the path within the oc-mirror cache directory to the local cache repositories.
+	cacheRepositoriesSubdir = "docker/registry/v2/repositories"
+
+	// tarRepositoriesPath is the path prefix for OCI repositories inside a tar archive.
+	tarRepositoriesPath = "docker/registry/v2/repositories/"
+)
+
+// TODO: Remove these structs once we move the integration tests into the oc-mirror repo
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/helpers_test.go:51]: This PR is the move but it is out of scope for now.

File: tests/integration/go.mod [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+	github.com/Microsoft/hcsshim v0.12.9 // indirect
+	github.com/antlr4-go/antlr/v4 v4.13.0 // indirect
+	github.com/asaskevich/govalidator v0.0.0-20230301143203-a9d515a09cc2 // indirect
+	github.com/beorn7/perks v1.0.1 // indirect
+	github.com/blang/semver/v4 v4.0.0 // indirect
+	github.com/bshuster-repo/logrus-logstash-hook v1.1.0 // indirect
+	github.com/cenkalti/backoff/v5 v5.0.3 // indirect
+	github.com/cespare/xxhash/v2 v2.3.0 // indirect
+	github.com/containerd/cgroups/v3 v3.0.3 // indirect
+	github.com/containerd/containerd v1.7.25 // indirect
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/go.mod:32]: solved

File: tests/integration/image-builders/release/Makefile [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+	go clean ./...
+
+container:
+	podman build -t  quay.io/oc-mirror/release/test-image:v0.0.1 -f containerfile-rhel9
+
+digest:
+	podman inspect quay.io/oc-mirror/release/test-image:v0.0.1 | jq '.[].Digest'
+
+push:
+	podman push quay.io/oc-mirror/release/test-image:v0.0.1 --authfile=/home/lzuccarelli/.docker/lmz-quay.json
```
[the remark this answers is not in the document]
Comment (adolfo-ab) [reply tests/integration/image-builders/release/Makefile]: Fixed

**Review by r4f4 — COMMENTED** [review]
/lgtm

**Review by aguidirh — APPROVED** [review]
Thanks @adolfo-ab 🎉

**Comment by aguidirh** [comment]
/verified bypass

Infra changes.
